### PR TITLE
Add tf/REPS

### PR DESCRIPTION
**Pull request by CatherineSue**
Add Relative Entropy Policy Search to garage/tf.
It is the same as garage theano/REPS except for RNN policy support. The
RNN support will be added after the RNN primitives in tf have tested.

Close: #363

### CODE REVIEW

File: garage/tf/algos/reps.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+    def _build_inputs(self):
+        """Decalre graph inputs variables."""
+        observation_space = self.policy.observation_space
+        action_space = self.policy.action_space
+        policy_dist = self.policy.distribution
+
+        with tf.name_scope("inputs"):
+            obs_var = observation_space.new_tensor_variable(
+                name="obs", extra_dims=2)
```
Comment (ryanjulian) [inline garage/tf/algos/reps.py]: i know YAPF won't force you to, but this block is probably more readable if each of these arguments are split into different lines

```python
valid_var = tensor_utils.new_tensor(
    name="valid",
    ndim=2,
    dtype=tf.float32)
```
...or similar

Comment (CatherineSue) [reply garage/tf/algos/reps.py]: em, but YAPF will force me to split into different lines
```
--- garage/tf/algos/reps.py	(original)
+++ garage/tf/algos/reps.py	(reformatted)
@@ -163,36 +163,22 @@
 
         with tf.name_scope("inputs"):
             obs_var = observation_space.new_tensor_variable(
-                name="obs",
-                extra_dims=2)
+                name="obs", extra_dims=2)

```

Comment (ryanjulian) [reply garage/tf/algos/reps.py]: oh i see

i think you can do 
```python
valid_var = tensor_utils.new_tensor(
    name="valid",
    ndim=2,
    dtype=tf.float32)  # yapf: disable
```

File: garage/tf/algos/reps.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+            return self.f_dual(*dual_opt_input_values)
+
+        def eval_dual_grad(x):
+            self.param_eta = x[0]
+            self.param_v = x[1:]
+            dual_opt_input_values = self._dual_opt_input_values(samples_data)
+            grad = self.f_dual_grad(*dual_opt_input_values)
+            eta_grad = np.float(grad[0])
+            v_grad = grad[1]
+            return np.hstack([eta_grad, v_grad])
```
Comment (ahtsan) [inline garage/tf/algos/reps.py]: Should we avoid creating inline functions?

File: tests/garage/tf/algos/test_reps.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            baseline=baseline,
+            batch_size=4000,
+            max_path_length=100,
+            n_itr=10,
+            discount=0.99,
+            max_kl_step=1e6,
+            plot=False)
+
+        last_avg_ret = algo.train(sess=self.sess)
+        assert last_avg_ret > 5
```
Comment (ahtsan) [inline tests/garage/tf/algos/test_reps.py]: Should the average return be higher to make sure it works?

File: garage/tf/algos/reps.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    def _build_inputs(self):
+        """Decalre graph inputs variables."""
+        observation_space = self.policy.observation_space
+        action_space = self.policy.action_space
+        policy_dist = self.policy.distribution
+
+        with tf.name_scope("inputs"):
+            obs_var = observation_space.new_tensor_variable(
+                name="obs",
+                extra_dims=2)   # yapf: disable
```
Comment (gonzaiva) [inline garage/tf/algos/reps.py]: I suppose these extra dims are for the batch sizes, but could we make that clearer by defining some constant instead of having these hard coded numbers.

File: garage/tf/algos/reps.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+                    name=k)
+                for k, shape in self.policy.state_info_specs
+            }   # yapf: disable
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k]
+                for k in self.policy.state_info_keys
+            ]   # yapf: disable
+
+            policy_old_dist_info_vars = {
+                k: tf.placeholder(
```
Comment (gonzaiva) [inline garage/tf/algos/reps.py]: We could call [new_tensor](https://github.com/rlworkgroup/garage/blob/1c4eebe59e4609be1a890c968c293f2d592ed3ad/garage/tf/misc/tensor_utils.py#L72) here.

File: garage/tf/regressors/gaussian_conv_regressor.py [hunk 6]
```diff
@@ -86,7 +86,7 @@ def __init__(self,
                             np.zeros(output_dim),
                             np.full(output_dim, np.log(init_std))
                         ],
-                        axis=0)
+                        axis=0)  # yapf: disable
```
Comment (gonzaiva) [inline garage/tf/regressors/gaussian_conv_regressor.py]: Is this due to a new version of yapf?

File: garage/tf/algos/reps.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+
+        return policy_loss_inputs, policy_opt_inputs, dual_opt_inputs
+
+    def _build_policy_loss(self, i):
+        """Initialize policy loss complie function based on inputs i."""
+        pol_dist = self.policy.distribution
+        is_recurrent = self.policy.recurrent
+
+        # Initialize dual params
+        self.param_eta = 15.
```
Comment (gonzaiva) [inline garage/tf/algos/reps.py]: Is there any particular reason why eta is 15?

File: garage/tf/algos/reps.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        )
+
+        return flatten_inputs(policy_opt_input_values)
+
+    def _features(self, samples_data):
+        """Get valid view features based on samples data."""
+        paths = samples_data["paths"]
+        feat_diff = []
+        for path in paths:
+            o = np.clip(path["observations"], -10, 10)
```
Comment (gonzaiva) [inline garage/tf/algos/reps.py]: Why is clipping min/max set to -10/10?

File: garage/tf/algos/reps.py (hunk 2 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: We should. But I don't know any simple way to make `eval_dual` and `eval_dual_grad` to be member functions or helper functions outside the class.

File: tests/garage/tf/algos/test_reps.py (hunk 3 above)
Comment (CatherineSue) [reply tests/garage/tf/algos/test_reps.py]: This is low because REPS is highly unstable and flaky.

File: garage/tf/algos/reps.py (hunk 5 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: em,,, we need to set `shape=[None] * 2 + list(shape)` here, but `new_tensor` returns a shape of `[None] * ndim`.

File: garage/tf/regressors/gaussian_conv_regressor.py (hunk 6 above)
Comment (CatherineSue) [reply garage/tf/regressors/gaussian_conv_regressor.py]: I think so but I don't remember.

File: garage/tf/algos/reps.py (hunk 5 above)
Comment (gonzaiva) [reply garage/tf/algos/reps.py]: oh, you're right, my bad

File: garage/tf/regressors/gaussian_conv_regressor.py (hunk 6 above)
Comment (gonzaiva) [reply garage/tf/regressors/gaussian_conv_regressor.py]: Okay, I just thought it's weird that nothing changed in this file and yapf complained.

File: garage/tf/algos/reps.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+        logger.record_tabular("{}/LossAfter".format(self.policy.name),
+                              loss_after)
+        logger.record_tabular("{}/dLoss".format(self.policy.name),
+                              loss_before - loss_after)
+        logger.record_tabular("{}/KLBefore".format(self.policy.name),
+                              policy_kl_before)
+        logger.record_tabular("{}/KL".format(self.policy.name), policy_kl)
+
+    def _build_inputs(self):
+        """Decalre graph inputs variables."""
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: Good catch, thx!

File: garage/tf/algos/reps.py (hunk 4 above)
Comment (ryanjulian) [reply garage/tf/algos/reps.py]: this is pretty common around the TF algos. the 2 extra dimensions are the valid bit and time.

i agree that this is annoying, but it's best dealt with as a structural change in the TF tree (probably after Anson gets models in)--not just here in this algo.

File: garage/tf/algos/reps.py (hunk 7 above)
Comment (ryanjulian) [reply garage/tf/algos/reps.py]: we can only squeeze so much magic out of hyperparameters, unfortunately.

File: garage/tf/algos/reps.py (hunk 8 above)
Comment (ryanjulian) [reply garage/tf/algos/reps.py]: this is worth a comment. i'm pretty sure it's to limit the dynamic range of the observations.

File: garage/tf/algos/reps.py (hunk 2 above)
Comment (ryanjulian) [reply garage/tf/algos/reps.py]: in this case i think it's fine. it comes down do a design issue with the optimizers

File: garage/tf/algos/reps.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+
+    """
+
+    def __init__(self,
+                 epsilon=0.5,
+                 l2_reg_dual=0.,
+                 l2_reg_loss=0.,
+                 optimizer=LbfgsOptimizer,
+                 optimizer_args=dict(max_opt_itr=50),
+                 dual_optimizer=scipy.optimize.fmin_l_bfgs_b,
```
Comment (naeioi) [inline garage/tf/algos/reps.py]: It is feasible to use LbfgsOptimizer instead? Otherwise, we have a divergent interface for optimizer against the other one above.

File: garage/tf/algos/reps.py [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+        policy_state_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.state_info_keys
+        ]   # yapf: disable
+        policy_old_dist_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.distribution.dist_info_keys
+        ]
+
+        self.feat_diff = self._features(samples_data)
```
Comment (naeioi) [inline garage/tf/algos/reps.py]: Features will be recalculated every time when dual_optimizer search for eta and v. This is not necessary.

File: garage/tf/algos/reps.py [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+            i.flat.policy_state_info_vars,
+            name="policy_dist_info_flat")
+
+        policy_dist_info_valid = filter_valids_dict(
+            policy_dist_info_flat,
+            i.flat.valid_var,
+            name="policy_dist_info_valid")
+
+        with tf.name_scope("bellman_error"):
+            delta_v = i.valid.reward_var + tf.tensordot(
```
[the remark this answers is not in the document]
Comment (zhanpenghe) [reply garage/tf/algos/reps.py]: Based on my understanding, this should be something like `r(s, a) + E(Phi * Theta) - Phi_s * Theta`.

File: garage/tf/algos/reps.py (hunk 10 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: No. I did find it is possible

File: garage/tf/algos/reps.py (hunk 11 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: I remember I put it here for a reason. But I will check if it is unnecessary here. Thanks

Comment (CatherineSue) [reply garage/tf/algos/reps.py]: `feat_diff` is used to calculate the bellman error. This is required here.

File: garage/tf/algos/reps.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: I think the above formula is true for REPS. Here the algorithm is policy iteration with REPS, which uses Table 2 in the paper.

File: garage/tf/algos/reps.py (hunk 8 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: I couldn't find how to compute features in the paper, so this method is ported from theano/algos/reps.py.

Comment (CatherineSue) [reply garage/tf/algos/reps.py]: I agree this is to limit the range of the observations. I don't know why they do it here instead of the sampler, so that we could normalize the observations, like Zhanpeng proposed. Another constant param is [100.0 in here](https://github.com/rlworkgroup/garage/pull/471/files#diff-1bc9d53a38cba8890b10755ab1366305R435). Is it related to max_path_length or max_reward?

File: garage/tf/algos/reps.py (hunk 11 above)
Comment (CatherineSue) [reply garage/tf/algos/reps.py]: I've moved the line out of this function.
